packages/pubsub/src/subscription.js

```diff
@@ -592,150 +395,315 @@ Subscription.prototype.decorateMessage_ = function(message) {
+      if (err) {
+        self.emit('error', err);
+      } else {
+        self.inventory_.nack = [];
+      }
+    });
+  }
+};
+
+
+/**
  * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this
+ *     request.
+ * @param {object} callback.apiResponse - Raw API response.
+ */
+Subscription.prototype.getMetadata = function(gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
+
+  var reqOpts = {
+    subscription: this.name
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'getSubscription',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
+
+/**
  *
- * @example
- * //-
- * // Pull all available messages.
- * //-
- * subscription.pull(function(err, messages) {
- *   // messages = [
- *   //   {
- *   //     ackId: '',     // ID used to acknowledge its receival.
- *   //     id: '',        // Unique message ID.
- *   //     data: '',      // Contents of the message.
- *   //     attributes: {} // Attributes of the message.
- *   //
- *   //     Helper functions:
- *   //     ack(callback): // Ack the message.
- *   //     skip():        // Free up 1 slot on the sub's maxInProgress value.
- *   //   },
- *   //   // ...
- *   // ]
- * });
+ */
+Subscription.prototype.hasMaxMessages_ = function() {
+  return this.inventory_.lease.length >= this.flowControl.maxMessages ||
+    this.inventory_.bytes >= this.flowControl.maxBytes;
+};
+
+/**
  *
- * //-
- * // Pull a single message.
- * //-
- * var opts = {
- *   maxResults: 1
- * };
+ */
+Subscription.prototype.leaseMessage_ = function(message) {
+  this.inventory_.lease.push(message.ackId);
+  this.inventory_.bytes += message.data.length;
+  this.setLeaseTimeout_();
+
+  return message;
+};
+
+/**
+ * Begin listening for events on the subscription. This method keeps track of
+ * how many message listeners are assigned, and then removed, making sure
+ * polling is handled automatically.
  *
- * subscription.pull(opts, function(err, messages, apiResponse) {});
+ * As long as there is one active message listener, the connection is open. As
+ * soon as there are no more message listeners, the connection is closed.
  *
- * //-
- * // If the callback is omitted, we'll return a Promise.
- * //-
- * subscription.pull(opts).then(function(data) {
- *   var messages = data[0];
- *   var apiResponse = data[1];
- * });
+ * @private
+ *
+ * @example
+ * subscription.listenForEvents_();
  */
-Subscription.prototype.pull = function(options, callback) {
+Subscription.prototype.listenForEvents_ = function() {
   var self = this;
-  var MAX_EVENTS_LIMIT = 1000;
 
-  if (!callback) {
-    callback = options;
-    options = {};
-  }
+  this.on('newListener', function(event) {
+    if (event === 'message') {
+      self.messageListeners++;
 
-  if (!is.number(options.maxResults)) {
-    options.maxResults = MAX_EVENTS_LIMIT;
-  }
+      if (!self.connection) {
```

Comment:
> Do you want to reopen the connection in the case that this.userClosed_ == true? 

I believe we do, if the user closed the connection but then added another `message` listener, I think we can only assume that they want to listen for messages again.

> And if you do, maybe you should set this.userClosed to false?

Yes! Good call.